source/uwp/Renderer/lib/AdaptiveActionRendererRegistration.cpp

```diff
@@ -0,0 +1,44 @@
+#include "pch.h"
+#include "AdaptiveActionRendererRegistration.h"
+#include "Util.h"
+
+using namespace Microsoft::WRL;
+using namespace ABI::AdaptiveNamespace;
+using namespace ABI::Windows::UI;
+
+namespace AdaptiveNamespace
+{
+    AdaptiveActionRendererRegistration::AdaptiveActionRendererRegistration() {}
+
+    HRESULT AdaptiveActionRendererRegistration::RuntimeClassInitialize() noexcept try
+    {
+        m_registration = std::make_shared<RegistrationMap>();
+        return S_OK;
+    }
+    CATCH_RETURN;
+
+    HRESULT AdaptiveActionRendererRegistration::Set(_In_ HSTRING type, _In_ IAdaptiveActionRenderer* renderer)
+    {
+        ComPtr<IAdaptiveActionRenderer> localRenderer(renderer);
+        (*m_registration)[HStringToUTF8(type)] = localRenderer;
+
+        return S_OK;
+    }
+
+    HRESULT AdaptiveActionRendererRegistration::Get(_In_ HSTRING type, _COM_Outptr_ IAdaptiveActionRenderer** result)
+    {
+        *result = nullptr;
+        RegistrationMap::iterator found = m_registration->find(HStringToUTF8(type));
+        if (found != m_registration->end())
+        {
+            found->second.CopyTo(result);
+        }
+        return S_OK;
+    }
+
+    HRESULT AdaptiveActionRendererRegistration::Remove(_In_ HSTRING type)
+    {
+        m_registration->erase(HStringToUTF8(type));
```

Comment:
>erase [](start = 24, length = 5)

will this throw if `type` isn't already registered?

Reply:
Looks like no. This is ultimately an unordered_map, which according to the documentation doesn't throw on not found. Just confirmed by adding a check to the test case.

---
In reply to: [261747139](https://github.com/Microsoft/AdaptiveCards/pull/2494#discussion_r261747139) [](ancestors = 261747139)